### ISSUE TITLE
[SharpDX.DXGI] Adapter.Description.Luid always return zero.

### ISSUE THREAD

**Issue by andrewst** [sharpdx/SharpDX#475]
I testing:
1.Windows 8.1 AMD Radeon HD 6900;
2.Windows 7 SP1 NVidia GT640;

**Comment by abenedik** [comment]
I have found a fix for that problem, but I cannot apply it because it requires to change the autogenerated code (AdapterDescription and AdapterDescription1).

The problem is that in x86 code the memory layout of the unmanaged AdapterDescription (and AdapterDescription1) is different from memory layout of managed AdapterDescription - or more precisely different from the __Native struct inside AdapterDescription.

In x86 the unmanaged layout looks like this (get with calling Adapter.GetDesc to fill byte array):
0 ... 256                                         Description
256   DE 10 00 00  C2 13 00 00     VendorId, DeviceId
264   43 10 08 85  A1 00 00 00      SubsystemId, Revision
272   00 00 00 C0  00 00 00 00      DedicatedVideoMemory, DedicatedSystemMemory
280   00 00 FF 3F  DF 97 00 00     SharedSystemMemory, Luid (low)
288   00 00 00 00  00 00 00 00      Luid (high), Flags

The Luid field is long in AdapterDescription. But in the table above it is written in two lines because it  is not aligned to 8 bytes but starts at 284 address.

In managed __Native struct the Luid is alligned to 8 bytes and starts at 288. Because flags are usually 0, the Luid is also almost always 0 (for me this is always the case). The problem is that the low 4 bytes are always lost.

This makes the Luid useless to compare adapters.

In x64 this problem does not occur, because DedicatedVideoMemory, DedicatedSystemMemory and SharedSystemMemory are all long (PointerSize) and therefore the Luid also aligns correctly.


Because the AdapterDescription and AdapterDescription1 are autogenerated I do not know how to fix that.

If that would not be the case, I would change the code as follows (I used DotPeak to get the decompiled code):

1) In _Native struct change the:

public long Luid;

into:

public uint Luid1, Luid2; // long Luid is split into 2 UInt32 to align correctly with native struct

2) In __MarshalFrom change:

this.Luid = @ref.Luid;

into:

this.Luid = ((long)@ref.Luid2 << 32) + @ref.Luid1;

3) Change __MarshalTo:

@ref.Luid = this.Luid;

into:

@ref.Luid2 = (uint)(this.Luid & 0xFFFFFFFF);
@ref.Luid1 = (uint)(this.Luid >> 32);


Maybe there is also an attribute that can be used to prevent wrong long alignment.

I hope this will give you enough info to fix the bug.
